src/mir.rs

```diff
@@ -675,6 +674,109 @@ impl MirFunction {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct MirForeignFunction {
+    name: Identifier,
+    foreign_name: Identifier,
+    args: Vec<(Identifier, MirType)>,
+    return_type: MirType,
+}
+
+impl MirForeignFunction {
+    pub fn new(
+        name: Identifier,
+        foreign_name: Identifier,
+        args: Vec<(Identifier, MirType)>,
+        return_type: MirType,
+    ) -> Self {
+        Self {
+            name,
+            foreign_name,
+            args,
+            return_type,
+        }
+    }
+
+    /// Declare this function to the compiler WITHOUT assembling it
+    fn declare(&self, funcs: &mut BTreeMap<Identifier, MirFunction>) -> Result<(), MirError> {
+        // Check if the function has already been declared
+        if funcs.contains_key(&self.name) {
+            Err(MirError::FunctionRedefined(self.get_name()))
+        } else {
+            // NOTE: because these declarations are only used for type checking and I dont want to rewrite them to much, we just define it as an oak function
+            funcs.insert(self.get_name(), MirFunction::new(
+                self.name.clone(),
+                self.args.clone(),
+                self.return_type.clone(),
+                vec![],
+            ));
```

Comment:
This is a bit questionable. Because `funcs` is only used for type checking this code won't give any issues. Changing this to the proper `MirForeignFunction` would require `funcs` and all its usages to be refactored to an `enum` to represent `MirFunction`'s and `MirForeignFunction`'s.